src/server/pkg/collection/transaction.go

```diff
@@ -281,3 +319,51 @@ func respToValue(key string, resp *v3.GetResponse) (string, error) {
 	}
 	return string(resp.Kvs[0].Value), nil
 }
+
+// fetchTTL contains the essential implementation of TTL().
+//
+// Note that 'iface' should either be the receiver 's' or a containing
+// 'stmSerializeable'--the only reason 'iface' is passed as a separate argument
+// is because fetchTTL calls iface.fetch(), and the implementation of 'fetch' is
+// different for stm and stmSerializeable. Passing the interface ensures the
+// correct version of fetch() is called
+func (s *stm) fetchTTL(iface STM, key string) (int64, error) {
+	// check wset cache
+	if wv, ok := s.wset[key]; ok {
+		return wv.ttl, nil
+	}
+
+	// Read ttl through s.ttlset cache
+	if ttl, ok := s.ttlset[key]; ok {
+		return ttl, nil
+	}
+
+	// Read kv and lease ID, and cache new TTL
+	getResp := iface.fetch(key) // *** call correct implementation of fetch()
+	if len(getResp.Kvs) == 0 {
+		panic("err not found: " + key)
+		return 0, ErrNotFound{Key: key}
+	}
+	leaseID := v3.LeaseID(getResp.Kvs[0].Lease) // type(kv.Lease) != v3.LeaseID?
+	if leaseID == 0 {
+		s.ttlset[key] = 0 // 0 is default value, but now 'ok' will be true on check
+		return 0, nil
+	}
+	leaseResp, err := s.client.TimeToLive(s.ctx, leaseID)
+	if err != nil {
+		panic(stmError{err})
```

Comment:
Ah, this is a little bit of a hack, but `Get` and `fetch` do the same thing. Basically, if you look at the implementation of `runSTM`, if one of these calls panics with an `stmError`, then `runSTM` recovers from the panic and passes the error to  the caller.

Basically since we invoke transactions like this:
```
col.NewSTM(ctx, etcdClient, func(stm STM) error {
  ...
  myCol.TT(key) // might break such that we can't commit the txn, but how to communicate that to NewSTM?
})
```
the way this communication happens is `NewSTM` calls this function, which calls `myCol.TTL(key)`. Then `myCol.TTL(key) panics, `NewSTM` calls `recover()`, and then retries the whole transaction without committing.

Reply:
Got it. Makes sense :+1: